ros-planning/moveit2
# Acceleration Limited Smoothing Plugin for Servo

### Description
This PR adds a new plugin called `AccelerationLimitedPlugin` under the `online_signal_smoothing::SmoothingBaseClass`  interface. The purpose of the plugin is to prevent the robot's acceleration limits from being violated by instantaneous changes to the servo command topics.  The plugin is not meant to act as a reference tracking controller, rather it's designed to be a layer between the hardware and command interface that prevents infeasible motions. 

There are three cases considered for acceleration-limiting illustrated in the following figures:
![Screenshot from 2024-01-12 16-45-36](https://github.com/ros-planning/moveit2/assets/28868750/ef78135a-030f-4116-a74a-4bafb23af63e)
In the figures, the red arrows show the displacement that will occur given the current velocity. The blue line shows the displacement between the current position and the desired position. The black dashed line shows the maximum possible displacements that are within the acceleration limits. The green line shows the acceleration commands that will be used by this acceleration-limiting plugin.  
 
- Figure A: The desired position is within the acceleration limits. The next commanded point will be exactly the desired point.
- Figure B: The line between the current position and the desired position intersects the acceleration limits, but the reference position is not within the bounds. The next commanded point will be the point on the displacement line that is closest to the reference.    
- Figure C: Neither the displacement line intersects the acceleration limits nor does the reference point lie within the limits. In this case, the next commanded point will be the one that minimizes the robot's velocity while maintaining its direction.

Case B can be solved via linear or quadratic program, hence I added OSQP as a dependency. This approach is compared to a more naive approach (where the required acceleration is scaled to avoid any limits from being violated) on a square tracing servoing motion (Left is this plugin).
 
![Screenshot from 2024-01-12 16-45-56](https://github.com/ros-planning/moveit2/assets/28868750/08c21f0c-4fd6-4e9e-baca-bac4a1d6a1f1)

The same motions are also shown as functions of time in the following plots. Importantly, the velocity of the left plot goes to zero in order to take the 90-degree turn.

![Screenshot from 2024-01-12 16-46-08](https://github.com/ros-planning/moveit2/assets/28868750/36b99680-2c22-488d-bf86-c7923cda94fa)

### Testing instructions: 
1. Launch `demo_ros_api.launch.py` from `moveit_serov`
2. Run `demo_ros_api.launch.py` from `servo_keyboard_input`
3. Use arrow keys to move the robot. You should see smoother motion compared to without the plugin. 

The update to the ros2_control usage may have broken this example. So to test, you will have to revert that commit.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit2/blob/main/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
